src/index.ts

```diff
@@ -287,54 +286,76 @@ const extension: JupyterFrontEndPlugin<void> = {
+      const toBoolean = (value: any): boolean => {
+        if (typeof value === 'boolean') return value;
+        if (typeof value === 'string') {
+          const lowercased = value.toLowerCase().trim();
+          return lowercased === 'true' || lowercased === '1';
+        }
+        return false;
+      };
+
+      // Convert configuration values to boolean
+      const enableBigQuery = toBoolean(bqFeature.enable_bigquery_integration);
+      const enableCloudStorage = toBoolean(
+        bqFeature.enable_cloud_storage_integration
+      );
+      const enableMetastore = toBoolean(bqFeature.enable_metastore_integration);
+      if (!enableMetastore) {
         // Preview was disabled, tear everything down.
         panelDpms?.dispose();
-        panelDatasetExplorer?.dispose();
+        panelDpms = undefined;
+        console.log('dispose');
+      }
+      if (!enableCloudStorage) {
         panelGcs?.dispose();
         gcsDrive?.dispose();
-        panelDpms = undefined;
-        panelDatasetExplorer = undefined;
         panelGcs = undefined;
         gcsDrive = undefined;
-      } else {
-        // Preview was enabled, (re)create DPMS and GCS.
-        if (!panelDpms && !panelGcs) {
-          panelDpms = new Panel();
-          panelDpms.id = 'dpms-tab';
-          panelDpms.title.caption = 'Dataset Explorer - DPMS';
-          panelDpms.addWidget(new dpmsWidget(app as JupyterLab, themeManager));
-          if (bqFeature.enable_bigquery_integration && !panelDatasetExplorer) {
-            panelDatasetExplorer = new Panel();
-            panelDatasetExplorer.id = 'dataset-explorer-tab';
-            panelDatasetExplorer.title.caption = 'Dataset Explorer - BigQuery';
-            panelDatasetExplorer.addWidget(
-              new BigQueryWidget(
-                app as JupyterLab,
-                settingRegistry as ISettingRegistry,
-                bqFeature.enable_bigquery_integration as boolean,
-                themeManager
-              )
-            );
-          }
-          panelGcs = new Panel();
-          panelGcs.id = 'GCS-bucket-tab';
-          panelGcs.title.caption = 'Google Cloud Storage';
-          gcsDrive = new GCSDrive();
-          documentManager.services.contents.addDrive(gcsDrive);
-          panelGcs.addWidget(
-            new GcsBrowserWidget(gcsDrive, factory as IFileBrowserFactory)
-          );
-          // Update the icons.
-          onThemeChanged();
-          app.shell.add(panelGcs, 'left', { rank: 1002 });
-          if (bqFeature.enable_bigquery_integration && panelDatasetExplorer) {
-            app.shell.add(panelDatasetExplorer, 'left', { rank: 1000 });
-          }
+      }
+      if (!enableBigQuery) {
+        panelDatasetExplorer?.dispose();
+        panelDatasetExplorer = undefined;
+      }
+      if (
+        bqFeature.enable_bigquery_integration ||
+        bqFeature.enable_cloud_storage_integration ||
+        bqFeature.enable_metastore_integration
+      ) {
```

Comment:
This way of structuring the logic is too brittle.

Instead, each feature should be structured of the form:

```
    if (enable<FEATURE>) {
        ...
    } else {
        panel<FEATURE>?.dispose();
        panel<FEATURE> = undefined;
    }
```

That way the code structure prevents the current bug where the GCSDrive is added to the document manager in cases where the feature is supposed to be disabled.

Reply:
<img width="637" alt="Screenshot 2024-12-16 at 10 32 29 AM" src="https://github.com/user-attachments/assets/88c88262-957c-4735-ac66-76db4e11950d" />
<img width="356" alt="Screenshot 2024-12-16 at 10 41 12 AM" src="https://github.com/user-attachments/assets/781bfbc7-d704-4d1e-b7ec-5c412f25322d" />

Using this structure results in random icons appearing when one or two features are enabled. However, when icons are added to the section all at once, they display correctly. Checking each flag individually and adding icons causes this inconsistency. To address this, we have adopted the approach of creating all widgets upfront and adding only the enabled ones. Please let us know if there are alternative approaches to ensure the correct icons are displayed.

Reply:
The issue with the icons must be happening somewhere else deeper in the codebase. Please find that and fix it.